Kustannusten seuranta

rckopteri opened the pull request:
Ensimmäisen vaiheen toteutus. 
Kustannusten seuranta-sivulle on toteutettu taulukko ja excel-raportti.

Toinen vaihe tulee myöhemmin kesällä.

Tässä ei ole otettu käyttöön vielä muuttunutta taulukko-komponenttia. Lisätään se toisessa vaiheessa tai kun se on valmis.

rckopteri committed:
Merge branch 'develop' into kustannusten-seuranta

# Conflicts:
#	src/cljs/harja/views/urakka/toteumat.cljs
#	tietokanta/src/main/resources/db/migration/V1_837__.sql

rckopteri committed:
Tee alustavasti näkymää kustannusten seurannalle

rckopteri committed:
Merge branch 'akilliset-UI' into kustannusten-seuranta

rckopteri committed:
Kustannusten seurantaan maksutyyppi

rckopteri committed:
Hae kustannuksille tehtävän nimi, jos se on annettu

Ja lisää päivämäärähaku

rckopteri committed:
Edistä ui lomaketta

rckopteri committed:
Merge branch 'develop' into kustannusten-seuranta

# Conflicts:
#	src/clj/harja/kyselyt/tehtavamaarat.sql
#	tietokanta/src/main/resources/db/migration/V1_838__.sql

rckopteri committed:
Merge branch 'akilliset-UI' into kustannusten-seuranta

# Conflicts:
#	src/cljs/harja/tiedot/urakka/urakka.cljs
#	tietokanta/src/main/resources/db/migration/V1_838__.sql

rckopteri committed:
Pieni keskeneräinen päivitys kustannusten seurantaan.

rckopteri committed:
Merge branch 'develop' into kustannusten-seuranta

rckopteri committed:
Merge branch 'develop' into kustannusten-seuranta

rckopteri committed:
Merge branch 'develop' into kustannusten-seuranta

rckopteri committed:
Aloita tekemään yksikkötestejä kustannusten seurannalle

rckopteri committed:
Esta null pointer poikkeus pakottamalla nolla budjetoituille summille

rckopteri committed:
Uudista kustannusten ryhmittelyä ui:lla

rckopteri committed:
Merge branch 'develop' into kustannusten-seuranta

rckopteri committed:
Päivitä hoidon johtoon liittyvää hakua

rckopteri committed:
Merge branch 'develop' into kustannusten-seuranta

rckopteri committed:
Päivitä kustannusten seurannan hoidon johdon sql hakua

rckopteri committed:
Merge branch 'develop' into kustannusten-seuranta

rckopteri committed:
Lisätään taulukkoa varten uuden tyylin mukaisia elementtejä

rckopteri committed:
Merge branch 'develop' into kustannusten-seuranta

rckopteri committed:
Tyylittele taulukkoon numeroiksi määritellyt elementit oikealle täsmättäviksi ja lisää punainen väri, jos luku on negatiivinen

rckopteri committed:
Poista turhaksi jäänyt toimenpiteiden haku kustannusten seurannan backendiltä

rckopteri committed:
Kustannusten seurannan tietokantahakujen päivitys

rckopteri committed:
Ryhmittele kustannukset pääryhmittäin ja näytä ne käyttöliittymässä jäsennetymmin

rckopteri committed:
Lisää hankintakustannusten budjetoinnille yksikkötesti

rckopteri committed:
Merge branch 'develop' into kustannusten-seuranta

rckopteri committed:
Kustannusten seurannan pääryhmien väripalluran suurentaminen

Ja muita marginaalisia refactorointeja

rckopteri committed:
Sää kustannusten seurannan ulkoasua lähemmäksi tavoitetta

rckopteri committed:
Lisää toteutunut_tyo taulu

Johon siirretään kustannusarvioitu_tyo taulusta kuukauden viimeinen päivä toteutuneet kustannukset

rckopteri committed:
Tee montako-paivaa-valissa pvm funktio ja sen testit

rckopteri committed:
Muutamia muutoksia kustannusten seurantaan

rckopteri committed:
Kustannusarvoiden siirto toteutunut_tyo tauluun on valmis.

rckopteri committed:
UI päivityksiä kustannusten seurantaan

rckopteri committed:
Lisää toteutunut_tyo taulu kustannusarvioitu_tyo ja johto_ja_hallintokorvaus taulujen toteumille.

Siirretään jo kustannuksiksi laskettavat rivit yllä olevista tauluista uuteen tauluun.
Päivitetään kustannusten seurannan sql haut käyttämään uutta taulua.

rckopteri committed:
Indeksikorotusten vaikutukset kustannusseurannassa

Indeksikorotukset lisätty sekä hakuun että yksikkötesteihin

rckopteri committed:
Kustannusseurantaan lisätty tavoitehinta ja kattohinta

rckopteri committed:
Hieman ui säätöä

rckopteri committed:
Korjattu smoke -testeissä löydetyt asiat

rckopteri committed:
Lisätään kuukausifiltteri kustannusten seurantaan

rckopteri committed:
Lisää hover taulukolle

rckopteri committed:
Lisää excel vienti-nappi uille ja valmistele sen toiminnallisuutta

rckopteri committed:
Excel raportin lisäys kustannus seurantaan

rckopteri committed:
Poista toimimaton excel raporttitesti

rckopteri committed:
Lisää taustaväri hoverille

rckopteri committed:
Merge branch 'develop' into kustannusten-seuranta

rckopteri committed:
Poista toimimaton less koodi

rckopteri committed:
Poista kommentoitu funktio

rckopteri committed:
Back testien korjaus

rckopteri committed:
Siivoa kommentit pois tehtävän nimestä

rckopteri committed:
Main testin korjaus

rckopteri committed:
Ja vielä avaimesta yksi kaksoispiste pois

rckopteri committed:
Lisätöiden tehtävät kustannusten seurannan sivulle ja excelraporttiin

rckopteri committed:
Lisätöille yksi lisätesti

rckopteri committed:
Kustannusten seurantaan kattohinta ja negatiivisille summille ui muutoksia

rckopteri committed:
Kustannusten seuranta - jaottele rahavauksiin ja hankintoihin

rckopteri committed:
Lisätään hankinnat ja rahavaraukset excel raporttiin kustannusten seurannassa

rckopteri committed:
Merge branch 'develop' into kustannusten-seuranta

# Conflicts:
#	src/clj/harja/palvelin/main.clj
#	test/clj/harja/palvelin/main_test.clj
#	tietokanta/src/main/resources/db/migration/V1_848__.sql

jpluomajoki-solita commented on dev-resources/less/vayla/common.less:
```diff
@@ -21,6 +21,7 @@
   }
 
   // Väylän typography hommat. Nämä voi ottaa pois sitten, kun väylän tyylit on otettu kaikkialla käyttöön.
+  //TODO: Nämä pitäisi siirtää typography.less tiedostoon
```
Onko tarkoitus tehdä myöhemmin?

rckopteri replied:
Laitoin vain ylös, että joskus joku noheva osais siirtää nämä oikeaan paikkaan

jpluomajoki-solita commented on src/cljs/harja/views/urakka/kulut/mhu_kustannusten_seuranta.cljs:
```diff
@@ -0,0 +1,429 @@
+     (big/fmt (big/mul (big/->big 100) (big/div toteuma suunniteltu)) 2)]))
+
+; spekseistä laskettu
+(def leveydet {:caret-paaryhma "2%"
```
Onko tällaisille vakioille erikseen määriteltyä paikkaa, vai onko tapana määritellä ne siellä, missä niitä käytetään?

rckopteri replied:
Normaalistihan riittää, että header elementille annetaan leveydet. Tämä taulukko näyttää olevan vain yksi taulukko, mutta on todellisuudessa koottu useammasta taulukosta, niin on tarpeen erotella nämä leveydet.

jpluomajoki-solita commented on src/cljs/harja/views/urakka/kulut/mhu_kustannusten_seuranta.cljs:
```diff
@@ -0,0 +1,429 @@
+                                 (pvm/nyt)
+                                 (pvm/->pvm (str "30.09." (inc valittu-hoitokauden-alkuvuosi))))
+                               nil)]
+    [:div.col-xs-12 {:style {:padding-top "24px"}}
```
Pitäisikö style-tagit määritellä muuttujina tai classeihin wrapattuna? Ainakin tässä näyttää olevan jonkin verran toistoa.

rckopteri replied:
Ehdottomasti classeinä. Nämä inline tyylit on kauheita. Siivoan pois.

jpluomajoki-solita commented on src/cljs/harja/views/urakka/kulut/mhu_kustannusten_seuranta.cljs:
```diff
@@ -0,0 +1,429 @@
+                        (second valittu-kuukausi)
+                        (str (inc valittu-hoitokausi) "-09-30"))]
+    [:div.kustannusten-seuranta
+     [debug/debug app]
```
Debug-valikko pois

jpluomajoki-solita commented on src/cljs/harja/views/urakka/kulut/mhu_kustannusten_seuranta.cljs:
```diff
@@ -0,0 +1,429 @@
+        [yleiset/livi-pudotusvalikko {:valinta valittu-hoitokausi
+                                      :vayla-tyyli? true
+                                      :valitse-fn #(e! (kustannusten-seuranta-tiedot/->ValitseHoitokausi (:id @nav/valittu-urakka) %))
+                                      :format-fn #(str "1.10." % "-30.9." (inc %))
```
Olisiko siistimpi/selkeämpi jos päivämäärät olisi vakioita? Nyt ainakin vaati hieman mentaalivoimistelua, että ymmärsi mistä oli kyse.
```
 (let [hoitokausi-alku-pvm "1.10."
        hoitokausi-loppu-pvm "30.9."]
        ... 
        :format-fn #(str hoitokausi-alku-pvm % " -" hoitokausi-loppu-pvm  (inc %))
```

rckopteri replied:
En tiedä parantuko, mutta muutin nämä muuttujiksi

jpluomajoki-solita commented on src/cljs/harja/views/urakka/kulut/mhu_kustannusten_seuranta.cljs:
```diff
@@ -0,0 +1,429 @@
+      [:h4 "Hoitovuosi: " hoitovuosi-nro " (1.10." valittu-hoitokauden-alkuvuosi " - 09.30." (inc valittu-hoitokauden-alkuvuosi) ")"]
+      (when hoitovuotta-jaljella
+        [:span "Hoitovuotta on jäljellä " hoitovuotta-jaljella " päivää."])]
+     [:div.table-default {:style {:padding-top "24px"}}
```
Taulu ei ole alignattu, otsikot oikealla ja sisältö vasemmalla.

rckopteri replied:
check

jpluomajoki-solita commented on src/cljs/harja/views/urakka/kulut/mhu_kustannusten_seuranta.cljs:
```diff
@@ -0,0 +1,429 @@
+                         muodostetut-tehtavat))))
+      toimenpiteet)))
+
+(defn- kustannukset-taulukko [e! app rivit-paaryhmittain]
```
Melko vaikealukuinen kokonaisuus, saisikohan tätä palasteltua helpommin purtavaksi?

rckopteri committed:
Merge branch 'develop' into kustannusten-seuranta

jpluomajoki-solita commented on src/cljs/harja/tiedot/urakka/kulut/mhu_kustannusten_seuranta.cljs:
```diff
@@ -0,0 +1,130 @@
+    (viesti/nayta! "Kattohinnan ja tavoitteen haku epäonnistui!" :danger)
+    app)
+
+  ;; Vain yksi rivi voi olla avattuna kerralla, joten tallennetaan avain app-stateen tai poistetaan se, jos se oli jo valittuna
```
Miksi vain yksi voi olla auki?

rckopteri replied:
Hyvä kysymys. Vastaus on, että se oli yksinkertaisempi toteuttaa näin. Mutta kysynpä ui suunnittelijalta, että onko tarvetta erilaiselle toteutukselle.

jpluomajoki-solita commented on src/clj/harja/palvelin/ajastetut_tehtavat/kustannusarvoiden_toteumat.clj:
```diff
@@ -0,0 +1,61 @@
+        nyt-kuukausi (pvm/kuukausi nyt)
+        viimeinen-paiva (t/day (t/last-day-of-the-month nyt-vuosi nyt-kuukausi))
+        kuukauden-loppupaiva (pvm/luo-pvm nyt-vuosi nyt-kuukausi viimeinen-paiva)
+        onko-kuukauden-viimeinen-paiva? (or (= nyt kuukauden-loppupaiva) false)
```
turha or

jpluomajoki-solita commented on src/clj/harja/palvelin/ajastetut_tehtavat/kustannusarvoiden_toteumat.clj:
```diff
@@ -0,0 +1,61 @@
+  Oletus on lähtökohtaisesti niin, että jos yksikin siirto on tehty, kaikki siirrot on tehty."
+  [db vuosi kuukausi]
+  (let [toteutuneet-siirrot (q/hae-toteutuneiden-siirtojen-maara db {:kuukausi kuukausi :vuosi vuosi})]
+    (if (> toteutuneet-siirrot 0)
```
turha if

jpluomajoki-solita commented on src/clj/harja/palvelin/ajastetut_tehtavat/kustannusarvoiden_toteumat.clj:
```diff
@@ -0,0 +1,61 @@
+      (do
+        (q/siirra-kustannusarvoidut-tyot-toteutumiin! db {:kuukausi nyt-kuukausi
+                                                          :vuosi nyt-vuosi})
+        (println "Siirto valamis!"))
```
debuggauslogi?

jpluomajoki-solita commented on src/clj/harja/palvelin/palvelut/kustannusten_seuranta.clj:
```diff
@@ -0,0 +1,235 @@
+      (throw+ {:type virheet/+sisainen-kasittelyvirhe+
+               :virheet [{:koodi 400
+                          :viesti "Tuntematon hoitokauden-alkuvuosi."}]})
+      (let [_ (log/info "hae-urakan-kustannusten-seuranta-paaryhmittain :: tiedot " (pr-str tiedot))
```
Onko tämä logitus tarkoituksella täällä? Samanlaisia logeja myös muualla.

jpluomajoki-solita commented on src/clj/harja/palvelin/palvelut/kustannusten_seuranta.clj:
```diff
@@ -0,0 +1,235 @@
+            :erotus nil
+            :prosentti nil}]
+          (mapcat
+            (fn [l]
```
l -> lisatyo?

jpluomajoki-solita commented on src/cljc/harja/domain/kulut/kustannusten_seuranta.cljc:
```diff
@@ -0,0 +1,176 @@
+                           (summaa-paaryhman-toimenpiteet 1 jjhallinnan-toimenpiteet)
+                           (summaa-tehtavat hoidonjohdonpalkkiot 2)
+                           (summaa-tehtavat erillishankinnat 3))
+        ;; TODO: meander - Ryhmittelyyn sopiva kirjasto, tutustuppa
```
Onko edelleen relevantti kommentti?

jpluomajoki-solita reviewed (commented):
Alkuun ainakin vähän kommentteja. Muutamia debuggausasioita näyttää jääneen vielä koodiin ja taulukon alignointi on vielä pielessä, mutta muuten toiminnallisuus näyttää hyvältä. Katselen vielä myöhemmin lisää.

rckopteri committed:
Lisää hoidon johto ja hallintakorvaukset siirrettäviin töihin myös

rckopteri committed:
Korjaa exceliin prosenttilasku

rckopteri committed:
Päivitä tavoitehinta ja kattohinta, kun hoitokautta vaihdetaan

rckopteri committed:
Lisää kustannusten seurantaan mahdollinen lisätyö

Lisätyötä ei haettu oikein jos se oli kodistettu johto ja hallintakorvaukselle. Samalla lisätty yksikkötesti

rckopteri committed:
Lisää vielä boldausta exceliin

rckopteri committed:
Korjaa ui:lla hoitovuoden näkyminen ja taulukon kolumnien leveyksiä

rckopteri committed:
Lisää tehtävien yhteenvedoille summatiedot paremmin, jotta luettavuus paranee, niin excelissä kuin ui:lla

rckopteri committed:
Poista react varoituksia, kun :key atribuutti puuttuu

rckopteri committed:
Katselmointikommenttien korjauksia

rckopteri committed:
Merge branch 'develop' into kustannusten-seuranta

# Conflicts:
#	test/clj/harja/palvelin/main_test.clj
#	tietokanta/testidata/laskutusyhteenveto_mhu.sql

rckopteri committed:
Korjaa lasku_kohdistus tauluun maksuerätyyppi, kun äkillinen hoitotyö tai vahinkojen korjaus lisätään laskulle

rckopteri committed:
Jaottele äkilliset hoitotyöt ja vahinkojen korjaukset uuteen varaukset pääryhmään

rckopteri committed:
Lisää varaukset pääryhmä käyttöliittymään

rckopteri committed:
Päivitä tietokantascriptin numero

rckopteri committed:
Järjestä varaukset toimenpiteen mukaiseen järjestykseen.

rckopteri committed:
Refactoroi kustannusten seurannan fronttikoodia

rckopteri committed:
Grouppaa kustannusten seurannan toteutuneita tehtäviä yhteen

Ja poista turhia kolumneja kyselyistä

rckopteri committed:
Merge branch 'tehtavaryhma-ja-laskyht-korjaus' into kustannusten-seuranta

# Conflicts:
#	tietokanta/testidata/suunnitellut_tyot.sql

rckopteri committed:
Lisää bonukset kustannusten seurannan ui:le

rckopteri committed:
Merge branch 'VHAR-3387-toteutuneet-tyot' into kustannusten-seuranta

# Conflicts:
#	src/clj/harja/kyselyt/toteutunut_tyo.clj
#	src/clj/harja/kyselyt/toteutunut_tyo.sql
#	tietokanta/src/main/resources/db/migration/V1_850__.sql

rckopteri committed:
Refaktoroi koodia hieman

rckopteri committed:
Mahdollista useamman pääryhmän "aukaisu" kerralla

rckopteri committed:
Bonusten listaus tuotu kustannusten seurannan ui:lle

rckopteri committed:
Merge branch 'VHAR-3387-toteutuneet-tyot' into kustannusten-seuranta

rckopteri committed:
Merge branch 'VHAR-3387-toteutuneet-tyot' into kustannusten-seuranta

rckopteri committed:
Kustannusten seuranta - palkkojen siirto palkkojen alle toimistotarvikkeiden alta

rckopteri committed:
Kustannusten seurannan excelin luonti erotettu omaan tiedostoon

rckopteri committed:
Merge branch 'VHAR-3387-toteutuneet-tyot' into kustannusten-seuranta

# Conflicts:
#	src/clj/harja/kyselyt/toteutuneet_kustannukset.clj
#	src/clj/harja/kyselyt/toteutuneet_kustannukset.sql

rckopteri committed:
Merge branch 'VHAR-3387-toteutuneet-tyot' into kustannusten-seuranta

# Conflicts:
#	src/clj/harja/palvelin/main.clj

rckopteri committed:
Refactor

rckopteri committed:
Yksikkötestien korjaus

rckopteri committed:
Merge branch 'develop' into kustannusten-seuranta

# Conflicts:
#	tietokanta/src/main/resources/db/migration/V1_851__.sql

rckopteri committed:
Kustannusten seurantaan testi vielä bonus pääryhmälle

rckopteri committed:
48 instanssin summat ovat muuttuneet, koska testidata on korjattu ja nyt tuonne oikeasti tulee tehtäväryhmät mukaan

joiden nimet oli aiemmin kirjoitettu väärin ja niitä ei löydetty.

rckopteri committed:
Korjataan hankintakustannusten testit

rckopteri committed:
Merge branch 'develop' into kustannusten-seuranta

rckopteri committed:
Kustannusten seurantaan korjattu bugi vhar3538 - lisätyöt hoidonjohdonpalkkioissa

rckopteri committed:
Merge branch 'VHAR-3387-toteutuneet-tyot' into kustannusten-seuranta

# Conflicts:
#	test/clj/harja/palvelin/main_test.clj
#	tietokanta/src/main/resources/db/migration/V1_851__.sql
#	tietokanta/src/main/resources/db/migration/V1_856__.sql

rckopteri committed:
Merge branch 'develop' into kustannusten-seuranta

rckopteri committed:
Muuta erillishankinnat tehtäväryhmän nimi yksilöiväksi tunnisteeksi

Ja lisää sopimus mukaan yksikkötestien hakuun

rckopteri committed:
Korjaa maksuerätesti
